distsql: adjust/tighten span partitions to avoid scanning over interleaved rows

richardwu opened the pull request:
Ignore the first commit since that is addressed in #20726

This is a follow-up from #20235 where we adjusted the span during the initial planning phases to avoid scanning over interleaved children.

This patch does the same for spans that have been partitioned during distsql planning.

Release notes: none

cc: @jordanlewis for the backfill span adjusting

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20694)
<!-- Reviewable:end -->

jordanlewis reviewed (commented):
Backfill span adjustments LGTM!
<!-- review by re -->

RaduBerinde commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_physical_planner.go, line 642 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20694#-L0H9L8D0fzP7bVqwP2W:-L0H9L8D0fzP7bVqwP2X:b8qko1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fedf279a12e72db86ab8522dbe975ae7f47855a/pkg/sql/distsql_physical_planner.go#L642)):*
> ```Go
> }
> 
> func adjustPartitionsForInterleave(partitions []spanPartition, table *sqlbase.TableDescriptor, index *sqlbase.IndexDescriptor) ([]spanPartition, error) {
> ```

This needs a comment. It always narrows down the spans (never extends them) correct?

---

*[pkg/sql/sqlbase/table.go, line 2634 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20694#-L0H9pjdmSEwsl5bslmy:-L0H9pjdmSEwsl5bslmz:bsgxhp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fedf279a12e72db86ab8522dbe975ae7f47855a/pkg/sql/sqlbase/table.go#L2634)):*
> ```Go
> 
> 	// len(keyTokens) > nIndexTokens
> 	// Case 3: tightened child key or grandchild key
> ```

Seems like the comment needs updating.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20694)*
<!-- Sent from Reviewable.io -->

richardwu committed:
distsql: tighten spans to not include interleaved children after partitioning

Release note (performance improvement): Queries with tables that have
interleaved tables should be faster now.

richardwu commented:
Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/distsql_physical_planner.go, line 642 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20694#-L0H9L8D0fzP7bVqwP2W:-L0kSuJ8JypnjtNlKFRd:b-nr1drm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fedf279a12e72db86ab8522dbe975ae7f47855a/pkg/sql/distsql_physical_planner.go#L642)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This needs a comment. It always narrows down the spans (never extends them) correct?
</blockquote></details>

Yes that is correct. I've renamed it as `tighten...` and added an appropriate comment.

---

*[pkg/sql/sqlbase/table.go, line 2634 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20694#-L0H9pjdmSEwsl5bslmy:-L0kSzYHH759tRs_Hu2u:bvljomq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fedf279a12e72db86ab8522dbe975ae7f47855a/pkg/sql/sqlbase/table.go#L2634)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Seems like the comment needs updating.
</blockquote></details>

Updated this in #20726, which is the parent/first commit of this PR (ignore the first commit in this PR).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20694)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_physical_planner.go, line 671 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20694#-L0ou-tOB7tbq0xHdjzZ:-L0ou-tOB7tbq0xHdjz_:by064nu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb89264e3cdb49a3358344ce5ed5f6b4673e55f/pkg/sql/distsql_physical_planner.go#L671)):*
> ```Go
> 			// and span generation.
> 			// If the key can be decomposed using
> 			// DecomposeKeyTokens, then it has not been invoked
> ```

I don't understand what "it has not been invoked with DecomposeKeyTokens" means. Did you mean PrefixEnd? I'd use "generated by" instead of "invoked with" (here and above).

---

*[pkg/sql/distsql_plan_backfill.go, line 77 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20694#-L0otuqx331_ESlmCdbv:-L0otuqx331_ESlmCdbw:bvdy1n7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb89264e3cdb49a3358344ce5ed5f6b4673e55f/pkg/sql/distsql_plan_backfill.go#L77)):*
> ```Go
> 		return physicalPlan{}, err
> 	}
> 	// Since the spans are used to read/write to the primary index, we can
> ```

Clean this up before merging

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20694#-:-L0ouIRnYeBxVPkPyXjC:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
@jordanlewis @RaduBerinde wasn't this pushed over the finish line since last year?

jordanlewis commented:
Yes, this was accomplished in other ways. Thanks @richardwu! 👋
